src/anomalib/engine/engine.py

```diff
@@ -0,0 +1,189 @@
+        return self._trainer
+
+    def _setup_trainer(self, model: AnomalyModule):
+        """Instantiates the trainer based on the model parameters."""
+        if self._cache.needs_override(model) or self._trainer is None:
+            self._cache.update(model)
+            self._trainer = Trainer(**self._cache.params)
+            # Callbacks need to be setup later as they depend on default_root_dir from the trainer
+            self._setup_anomalib_callbacks()
+
+    def _setup_anomalib_callbacks(self):
+        """Setup callbacks for the trainer."""
+
+        _callbacks: list[Callback] = [_PostProcessorCallback()]
+        normalization_callback = get_normalization_callback(self.normalizer)
+        if normalization_callback is not None:
+            _callbacks.append(normalization_callback)
+
+        _callbacks.append(_ThresholdCallback(self.threshold))
+        _callbacks.append(_MetricsCallback(self.task, self.image_metric_names, self.pixel_metric_names))
+
+        if self.visualization is not None:
+            image_save_path = self.visualization.pop("image_save_path", None)
+            if image_save_path is None:
+                image_save_path = self.trainer.default_root_dir + "/images"
+            _callbacks += get_visualization_callbacks(
+                task=self.task, image_save_path=image_save_path, **self.visualization
+            )
+
+        self.trainer.callbacks = _CallbackConnector._reorder_callbacks(self.trainer.callbacks + _callbacks)
+
+    def fit(
+        self,
+        model: AnomalyModule,
+        train_dataloaders: TRAIN_DATALOADERS | LightningDataModule = None,
+        val_dataloaders: EVAL_DATALOADERS = None,
+        datamodule: LightningDataModule | None = None,
+        ckpt_path: str | None = None,
+    ) -> None:
+        self._setup_trainer(model)
+        self.trainer.fit(model, train_dataloaders, val_dataloaders, datamodule, ckpt_path)
+
+    def validate(
+        self,
+        model: AnomalyModule | None = None,
+        dataloaders: EVAL_DATALOADERS | LightningDataModule = None,
+        ckpt_path: str | None = None,
+        verbose: bool = True,
+        datamodule: LightningDataModule | None = None,
+    ) -> _EVALUATE_OUTPUT | None:
+        if model:
+            self._setup_trainer(model)
+        return self.trainer.validate(model, dataloaders, ckpt_path, verbose, datamodule)
+
+    def test(
+        self,
+        model: AnomalyModule | None = None,
+        dataloaders: EVAL_DATALOADERS | LightningDataModule = None,
+        ckpt_path: str | None = None,
+        verbose: bool = True,
+        datamodule: LightningDataModule | None = None,
+    ) -> _EVALUATE_OUTPUT:
+        if model:
+            self._setup_trainer(model)
+        return self.trainer.test(model, dataloaders, ckpt_path, verbose, datamodule)
+
+    def predict(
+        self,
+        model: AnomalyModule | None = None,
+        dataloaders: EVAL_DATALOADERS | LightningDataModule = None,
```

Comment:
```suggestion
        dataloaders: EVAL_DATALOADERS | LightningDataModule | None = None,
```

Reply:
I can add that. I guess since EVAL_DATALOADERS is of type `Any`, it didn't give any error